ed25519.go

```diff
@@ -0,0 +1,170 @@
+//go:build !cmd_go_bootstrap
+
+package openssl
+
+// #include "goopenssl.h"
+import "C"
+import (
+	"errors"
+	"strconv"
+	"sync"
+	"unsafe"
+)
+
+const (
+	// publicKeySize is the size, in bytes, of public keys as used in crypto/ed25519.
+	publicKeySizeEd25519 = 32
+	// privateKeySize is the size, in bytes, of private keys as used in crypto/ed25519.
+	privateKeySizeEd25519 = 64
+	// signatureSize is the size, in bytes, of signatures generated and verified by crypto/ed25519.
+	signatureSizeEd25519 = 64
+	// seedSize is the size, in bytes, of private key seeds. These are the private key representations used by RFC 8032.
+	seedSizeEd25519 = 32
+)
+
+// TODO: Add support for Ed25519ph and Ed25519ctx when OpenSSL supports them,
+// which will probably be in 3.2.0.
+
+var (
+	onceSupportsEd25519 sync.Once
+	supportsEd25519     bool
+)
+
+// SupportsEd25519 returns true if the current OpenSSL version supports
+// GenerateKeyEd25519, NewKeyFromSeedEd25519, SignEd25519 and VerifyEd25519.
+func SupportsEd25519() bool {
+	onceSupportsEd25519.Do(func() {
+		switch vMajor {
+		case 1:
+			supportsEd25519 = version1_1_1_or_above()
+		case 3:
+			name := C.CString("ED25519")
+			defer C.free(unsafe.Pointer(name))
+			sig := C.go_openssl_EVP_SIGNATURE_fetch(nil, name, nil)
+			if sig != nil {
+				C.go_openssl_EVP_SIGNATURE_free(sig)
+				supportsEd25519 = true
+			}
+		}
+	})
+	return supportsEd25519
+}
+
+func GenerateKeyEd25519() (pub, priv []byte, err error) {
+	pkey, err := generateEVPPKey(C.GO_EVP_PKEY_ED25519, 0, "")
+	if err != nil {
+		return nil, nil, err
+	}
+	defer C.go_openssl_EVP_PKEY_free(pkey)
+	priv = make([]byte, privateKeySizeEd25519)
+	err = extractPKEYPrivEd25519(pkey, priv)
+	if err != nil {
+		return nil, nil, err
+	}
+	pub = make([]byte, publicKeySizeEd25519)
+	copy(pub, priv[seedSizeEd25519:])
+	return pub, priv, nil
+}
+
+func NewKeyFromSeedEd25519(seed []byte) (priv []byte, err error) {
```

Comment:
I think it might be worth mentioning that "seed" is based on the naming that crypto/ed25519 uses, and that's also why the private key includes the public key inside it. It might be obvious enough to check that given the purpose of golang-fips, but it took me a bit because I was looking at the OpenSSL side first.

(Maybe simply copy the package doc from crypto/ed25519 to this file for reference?)